spring-rabbit/src/main/java/org/springframework/amqp/rabbit/listener/DirectMessageListenerContainer.java

```diff
@@ -0,0 +1,244 @@
+import org.springframework.util.Assert;
+
+import com.rabbitmq.client.AMQP.BasicProperties;
+import com.rabbitmq.client.Channel;
+import com.rabbitmq.client.DefaultConsumer;
+import com.rabbitmq.client.Envelope;
+
+/**
+ * The {@code SimpleMessageListenerContainer} is not so simple. Recent changes to the
+ * rabbitmq java client has facilitated a much simpler listener container that invokes the
+ * listener directly on the rabbit client consumer thread. There is no txSize property -
+ * each message is acked (or nacked) individually.
+ * @author Gary Russell
+ * @since 2.0
+ *
+ */
+public class DirectMessageListenerContainer extends AbstractMessageListenerContainer {
+
+	private final List<SimpleConsumer> consumers = new ArrayList<SimpleConsumer>();
+
+	private final MessagePropertiesConverter messagePropertiesConverter = new DefaultMessagePropertiesConverter();
+
+	private final ContainerDelegate delegate = new ContainerDelegate() {
+		@Override
+		public void invokeListener(Channel channel, Message message) throws Exception {
+			DirectMessageListenerContainer.super.invokeListener(channel, message);
+		}
+	};
+
+	private ContainerDelegate proxy = this.delegate;
+
+	private volatile TaskExecutor taskExecutor;
+
+	private volatile int consumersPerQueue = 1;
+
+	private volatile int prefetch = 1;
+
+	private volatile boolean defaultRequeueRejected = false;
+
+	private volatile long recoveryInterval;
+
+	/**
+	 * Create an instance with the provided connection factory.
+	 * @param connectionFactory the connection factory.
+	 */
+	public DirectMessageListenerContainer(ConnectionFactory connectionFactory) {
+		setConnectionFactory(connectionFactory);
+	}
+
+	@Override
+	public final void setConnectionFactory(ConnectionFactory connectionFactory) {
+		super.setConnectionFactory(connectionFactory);
+	}
+
+	/**
+	 * Set a task executor for the container - used to create the consumers not at
+	 * runtime.
+	 * @param taskExecutor the task excutor.
+	 */
+	public void setTaskExecutor(TaskExecutor taskExecutor) {
+		this.taskExecutor = taskExecutor;
+	}
+
+	/**
+	 * Each queue runs in its own consumer; set this property to create multiple
+	 * consumers for each queue.
+	 * @param consumersPerQueue the consumers per queue.
+	 */
+	public void setConsumersPerQueue(int consumersPerQueue) {
+		this.consumersPerQueue = consumersPerQueue;
+	}
+
+	/**
+	 * Set to false to requeue failed deliveries.
+	 * @param requeueRejected the requeue rejected.
+	 */
+	public void setDefaultRequeueRejected(boolean requeueRejected) {
+		this.defaultRequeueRejected = requeueRejected;
+	}
+
+	@Override
+	protected void doInitialize() throws Exception {
+		Assert.notNull(getQueueNames(), "queue(s) are required");
+		ContainerDelegate proxy = initializeProxy(this.delegate);
+		if (proxy != null) {
+			this.proxy = proxy;
+		}
+	}
+
+	@Override
+	protected void invokeListener(Channel channel, Message message) throws Exception {
+		this.proxy.invokeListener(channel, message);
+	}
```

Comment:
Looks like this method is fully equal to the same in the `SimpleMessageListenerContainer`.
We really can move it to the `AbstractMessageListenerContainer`.
Great work!